onnxruntime/core/providers/cuda/tensor/onehot.cu

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/cuda_common.h"
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "core/providers/cuda/tensor/onehot.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+template <typename in_type, typename out_type>
+__global__ void _OneHotImpl(
+    const in_type* indices_data,
+    const fast_divmod fdm_depth_suffix,
+    const fast_divmod fdm_suffix,
+    const int64_t depth_val,
+    const out_type on_value,
+    const out_type off_value,
+    out_type* output_data,
+    CUDA_LONG N) {
+  CALCULATE_ELEMENTWISE_INDEX_OR_EXIT(id, N);
+
+  int prefix_index, prefix_offset;
+  fdm_depth_suffix.divmod(id, prefix_index, prefix_offset);
+
+  int depth_index, suffix_index;
+  fdm_suffix.divmod(prefix_offset, depth_index, suffix_index);
+
+  CUDA_LONG indices_index = prefix_index * fdm_suffix.d_ + suffix_index;
+
+  // handle negative index
+  in_type adjusted_indice = (indices_data[indices_index] + depth_val) % depth_val;
```

Comment:
I don't know how % performs on gpu. do you think putting this in onehot.cc be better (just as CPU OneHot kernel did).

Reply:
indices_data lives in GPU memory. 
I think we have no choice but to handle the negative index in cuda kernel.

Reply:
you are right!